clang/lib/CIR/Lowering/ThroughMLIR/LowerCIRLoopToSCF.cpp

```diff
@@ -570,6 +570,81 @@ class CIRWhileOpLowering : public mlir::OpConversionPattern<cir::WhileOp> {
     }
   }
 
+  void rewriteBreak(mlir::scf::WhileOp whileOp,
+                    mlir::ConversionPatternRewriter &rewriter) const {
+    // Collect all BreakOp inside this while.
+    llvm::SmallVector<cir::BreakOp> breaks;
+    whileOp->walk([&](mlir::Operation *op) {
+      if (auto breakOp = dyn_cast<BreakOp>(op))
+        breaks.push_back(breakOp);
+    });
+
+    if (breaks.empty())
+      return;
+
+    for (auto breakOp : breaks) {
+      bool nested = false;
+      // When there is another loop between this WhileOp and the BreakOp,
+      // we should change that loop instead.
+      for (mlir::Operation *parent = breakOp->getParentOp(); parent != whileOp;
+           parent = parent->getParentOp()) {
+        if (isa<WhileOp>(parent)) {
+          nested = true;
+          break;
+        }
+      }
+      if (nested)
+        continue;
+
+      // Similar to the case of ContinueOp, when there is an `IfOp`,
+      // we need to take special care.
+      for (mlir::Operation *parent = breakOp->getParentOp(); parent != whileOp;
+           parent = parent->getParentOp()) {
+        if (auto ifOp = dyn_cast<cir::IfOp>(parent))
+          llvm_unreachable("NYI");
+      }
+
+      // Operations after this BreakOp has to be removed.
+      for (mlir::Operation *runner = breakOp->getNextNode(); runner;) {
```

Comment:
Why you need to remove this? Looks like we should just split the block at this point and create an unrecheable block (which should get later DCE'd by canonicalizer)?